### ISSUE TITLE
Generating swagger documentation creates large increase in unmanaged memory allocation that is not released

### ISSUE THREAD

**Issue by statler** [dotnet/aspnetcore#30278]
### Describe the bug
I am not sure if it is by design, but when generating a large swagger document, a large amount of unmanaged memory is allocated and not released - in my case 300mb, doubling my application's memory footprint. 

### To Reproduce
Generate a large swagger document (mine is >2Mb) during startup (I use NSwag). 

    services.AddSwaggerDocument()

It s described in a preliminary discussion with the NSwag developers here https://github.com/RicoSuter/NSwag/issues/3306 where it is confirmed that their resources are released. As per the discussion, it is possible that the memory allocation is benign, but it seems excessive and is problematic when trying to monitor memory resource utilization on azure.

- Normal start up to 10s
- Document load 25-35s
- At rest by 45s

![image](https://user-images.githubusercontent.com/337991/108431426-f66e0b80-7296-11eb-8e85-691acdae2702.png)

Without swagger generation

![image](https://user-images.githubusercontent.com/337991/108430963-4f896f80-7296-11eb-943b-2ef027b3e45c.png)

All of the managed memory observations are as expected.

### Exceptions (if any)
None

### Further technical details
- ASP.NET Core version 5
- VS 2019
- IIS Express debug

.NET SDK (reflecting any global.json):
 Version:   5.0.102
 Commit:    71365b4d42

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.18363
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\5.0.102\

Host (useful for support):
  Version: 5.0.2
  Commit:  cb5f173b96

.NET SDKs installed:
  2.1.801 [C:\Program Files\dotnet\sdk]
  2.1.802 [C:\Program Files\dotnet\sdk]
  2.2.104 [C:\Program Files\dotnet\sdk]
  5.0.101 [C:\Program Files\dotnet\sdk]
  5.0.102 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.13 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.24 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.13 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.24 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.24 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.11 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.2 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

**Comment by javiercn** [comment]
@statler thanks for contacting us.

The swagger library is not part of ASP.NET Core, it's a third-party package. I don't think we allocate unmanaged memory anywhere so I would suggest following up with the library authors?

@bradygaster can you point to the right place?

**Comment by statler** [comment]
@javiercn 

As per my ticket, I have already traced this with the vendor and their libraries properly clean up their resources. The reason the ticket is here is because this may well be associated with the API Explorer, which is leveraged by the swagger gen.

**Comment by javiercn** [comment]
@statler sorry about that, late working hours 😄.

I'm 99.99% sure we don't allocate any unmanaged memory in API explorer. I can't think of a way to track the unmanaged memory back to the managed component allocating it of the top of my head.

Can you share more details about the investigation you've done so far as well as a minimal repro so that we can take a look?

**Comment by javiercn** [comment]
@statler do you also see this behavior if you remove the swagger support and just exercise the code through API explorer?

**Comment by statler** [comment]
No problem @javiercn :)

How do I go about executing the API Explorer when I am not using MVC? This is a straight API with AddControllers()

            services.AddControllers(config =>
            {
                config.Filters.Add(new AuthorizeFilter(authorizePolicy));
                config.OutputFormatters.OfType<StringOutputFormatter>().Single().SupportedMediaTypes.Add("text/html");
            })

**Comment by javiercn** [comment]
@statler I'm not 100% sure what the NSwag and Swashbuckle use API explorer, but very likely they iterate over all the elements in the model.

I would suggest injecting api explorer into an API controller and iterating over all the objects in the model and then returning a simple response, like Ok and the number of operations or similar (haven't touch API explorer in a while, so I can't point right away to the right type).

**Comment by bradygaster** [comment]
AFAIK, Swashbuckle does use it. Adding @domaindrivendev here to confirm.

**Comment by statler** [comment]
After some more research, including articles such as https://github.com/dotnet/aspnetcore/issues/3409, it is possible that some of this is the opaqueness I feel when trying to understand what .Net Core does memory wise.

1. What is the end-game for the Unmanaged memory processes - is that memory ever collectable/reallocatable? I understand it is out of the bounds of the garbage collector, but does this mean that in the below case I have 300Mb of Dlls and other framework / lib that is just no longer available after startup? I have done my best to understand this but simply dont.
2. I get that garbage collection is [by default] greedy in server mode. I have a similar case to the link above where I have multiple app services running on the one app plan and to have this kind of memory consumption up until a critical point where it ends up going into garbage collection panic and everything falls over until it sorts itself out again, is not a great way to operate. Is there asp.net API specific information on running services in a shared enironment, including how to configure memory and GC for specific common scenarios?

![Nswag](https://user-images.githubusercontent.com/337991/108644514-65e83300-7503-11eb-89b1-e54c2d780d7f.png)

**Comment by javiercn** [comment]
@sebastienros @pranavkm do you have any thoughts here?

@statler I'm going on a limb here, but I believe some of that memory can be pages that were allocated as part of the memory increase and that have yet to be returned to the operating system.

**Comment by sebastienros** [comment]
First thing to do is to collect a trace that will show what objects are actually holding this memory. Since you are using dotmemory it should already be able to list all types. Otherwise perfview with the default settings will also report it.

**Comment by sebastienros** [comment]
Notes: 
- The GC collects when it feels the need to do so. If you have lots of available memory the OS won't feel any pressure and the GC might not be invoked when you wish it would. But it the nswag thread you seem to call GC.Collect() and it doesn't change the result, so in this case it would not be an issue with when the GC is invoked. Though even in that case, it would not be a bug, just by design.
- I remember that nswag is using dotliquid as a template engine. My experience with it is that it allocates a lot and might be the reason for the increased memory, even more if the document is big. Do you know if dotliquid could be used during the process you are describing?
- Maybe taking a memory dump and using windbg can tell us more about the native memory if dotmemory can't show that. But try perfview first.

Feel free to skype if you need some help taking the profile. (same username)

**Comment by statler** [comment]
Thanks Sebastienros - totally fine on the memory profiling, though based on some diagnostic logging I am trialling, I may well take you up on your skype offer in the coming days when I have a firmer idea of the base causes. I am now [relatively] comfortable that the issue isn't in the _managed_ memory with either API Explorer, NSwag or Swashbuckle. Simililarly I experience no issues when working in dev. If I call a GC on the app manually (I have a dummy controller I use for this), then managed memory baselines back to 70Mb. Everything I see in the Profile, i pretty much expect to be there.

My issues are when I get to Azure. What I can profile there is relatively limited, as Azure just gives total memory consumption, and that seems to mirror what I see with the Total Memory (Unmanaged + Managed) when in development and as shown in my profiling. If I leave my app running, it eventually just slows after a couple of days until it eventually gets super sluggish and can even not respond for 5-10 minutes. Then usually it seems to come good for a bit. Rinse repeat. The memory profile is consistently high (up around 60-70%) which mirrors Total memory. The app is essentially a large entity framework business app.

So, in summary - I understand what the app is doing managed memory wise, and GC returns it to where I expect it to be. I don't see any increase that GC doesnt resolve.

What I am left with is:
- that maybe GC is not occurring when needed - in the interim I am trialling with GC every request to test that theory.
- there is some problem affecting unmanaged memory.

It is the latter that I am inexperienced with and am trying to resolve. Any pointers welcome, in the meantime I continue with diagnostics and logging. Thanks again for everyone's help

**Comment by javiercn** [comment]
@statler another suggestion is that this likely helps confirm that it's just the GC not returning the pages (on purpose) to the OS. Checkout this [setting](https://docs.microsoft.com/en-us/dotnet/core/run-time-config/garbage-collector#retain-vm)

**Comment by statler** [comment]
Oh wow. That is so helpful

Thank you

**Comment by statler** [comment]
Just trying to get my head around Retain VM. As I understand it default is false. Wouldnt the default behaviour mean that the resources are returned to the system at the earliest convenience? W

**Comment by statler** [comment]
Sorry - didnt mean to close

**Comment by javiercn** [comment]
@statler did you get to try out our suggestion?

We don't think there is an issue in ASP.NET Core here and that this is just the expected behavior from the GC where it will keep the pages from the OS around in case it needs them.
